threedi_api_client/aio/files.py

```diff
@@ -0,0 +1,401 @@
+    start, stop, total = [
+        int(x) for x in CONTENT_RANGE_REGEXP.findall(content_range)[0]
+    ]
+    return response, total
+
+
+async def download_fileobj(
+    url: str,
+    fileobj,
+    chunk_size: int = 16777216,
+    timeout: float = 5.0,
+    connector: Optional[aiohttp.BaseConnector] = None,
+    retries: int = 3,
+    backoff_factor: float = 1.0,
+) -> int:
+    """Download a url to a file object using multiple requests.
+
+    It is assumed that the file server supports multipart downloads (range
+    requests).
+
+    Args:
+        url: The url to retrieve.
+        fileobj: The (binary) file object to write into, supporting async I/O.
+        chunk_size: The number of bytes per request. Default: 16MB.
+        timeout: The total timeout in seconds.
+        connector: An optional aiohttp connector to support connection pooling.
+            If not supplied, a default TCPConnector is instantiated with a pool
+            size (limit) of 4.
+        retries: Total number of retries per request.
+        backoff_factor: Multiplier for retry delay times (1, 2, 4, ...)
+
+    Returns:
+        The total number of downloaded bytes.
+
+    Raises:
+        openapi_client.exceptions.ApiException: raised on unexpected server
+            responses (HTTP status codes other than 206, 413, 429, 503)
+        aiohttp.ClientError: various low-level HTTP errors that persist
+            after retrying: connection errors, timeouts, decode errors,
+            invalid HTTP headers, payload too large (HTTP 413), too many
+            requests (HTTP 429), service unavailable (HTTP 503)
+    """
+    if connector is None:
+        connector = aiohttp.TCPConnector(limit=DEFAULT_CONN_LIMIT)
+
+    # Our strategy here is to download the first chunk, get the total file
+    # size from the header, and then parrellelize the rest of the chunks.
+    # We could get the total Content-Length from a HEAD request, but not all
+    # servers support that (e.g. Minio).
+    request_kwargs = {
+        "url": url,
+        "timeout": timeout,
+        "retries": retries,
+        "backoff_factor": backoff_factor,
+    }
+    async with aiohttp.ClientSession(connector=connector) as client:
+        # start with a single chunk to learn the total file size
+        response, file_size = await _download_request(
+            client, 0, chunk_size, **request_kwargs
+        )
+
+        # write to file
+        await fileobj.write(await response.read())
+
+        # return if the file is already completely downloaded
+        if file_size <= chunk_size:
+            return file_size
+
+        # create tasks for the rest of the chunks
+        tasks = [
+            asyncio.ensure_future(
+                _download_request(client, start, start + chunk_size, **request_kwargs)
+            )
+            for start in range(chunk_size, file_size, chunk_size)
+        ]
+
+        # write the result of the tasks to the file one by one
+        for task in tasks:
```

Comment:
I think you actually have to use `gather` to run the task sequence concurrently.
Something like this maybe?
```
task_results = await asyncio.gather(*tasks, return_exceptions=True)
for task_result in task_results:
    if isinstance(task_result, Exception):
        ...
    await fileobj.write(await task_result.read())

Reply:
The problem with `gather` is that at some point, you will have the complete file in memory. It should be streamed to disk ASAP else we get in trouble with large files.

Putting the write itself inside the `gather` is also not an option, because it gives no guarantees on the execution order of the tasks. So you might scamble the order of chunks.

This approach seems to work (`ensure_future` or `create_tasks` directly start executing)